tests/integration/ambient/waypoint_test.go

```diff
@@ -61,3 +67,131 @@ func TestWaypointStatus(t *testing.T) {
 			retry.UntilSuccessOrFail(t, check)
 		})
 }
+
+func TestWaypoint(t *testing.T) {
+	framework.
+		NewTest(t).
+		Features("traffic.ambient").
+		Run(func(t framework.TestContext) {
+			nsConfig, err := namespace.New(t, namespace.Config{
+				Prefix: "waypoint",
+				Inject: false,
+				Labels: map[string]string{
+					constants.DataplaneMode: "ambient",
+				},
+			})
+			if err != nil {
+				t.Fatal(err)
+			}
+
+			istioctl.NewOrFail(t, t, istioctl.Config{}).InvokeOrFail(t, []string{
+				"x",
+				"waypoint",
+				"apply",
+				"--namespace",
+				nsConfig.Name(),
+			})
+			retry.UntilSuccessOrFail(t, func() error {
+				fetch := kubetest.NewPodFetch(t.AllClusters()[0], nsConfig.Name(), constants.GatewayNameLabel+"="+"namespace")
+				if _, err := kubetest.CheckPodsAreReady(fetch); err != nil {
+					return fmt.Errorf("gateway is not ready: %v", err)
+				}
+				return nil
+			}, retry.Timeout(15*time.Second), retry.BackoffDelay(time.Millisecond*100))
+
+			saSet := []string{"sa1", "sa2", "sa3"}
+			for _, sa := range []string{"sa1", "sa2", "sa3"} {
+				istioctl.NewOrFail(t, t, istioctl.Config{}).InvokeOrFail(t, []string{
+					"x",
+					"waypoint",
+					"apply",
+					"--namespace",
+					nsConfig.Name(),
+					"--service-account",
+					sa,
+				})
+				retry.UntilSuccessOrFail(t, func() error {
+					fetch := kubetest.NewPodFetch(t.AllClusters()[0], nsConfig.Name(), constants.GatewayNameLabel+"="+sa)
+					if _, err := kubetest.CheckPodsAreReady(fetch); err != nil {
+						return fmt.Errorf("gateway is not ready: %v", err)
+					}
+					return nil
+				}, retry.Timeout(15*time.Second), retry.BackoffDelay(time.Millisecond*100))
+			}
+
+			output, _ := istioctl.NewOrFail(t, t, istioctl.Config{}).InvokeOrFail(t, []string{
+				"x",
+				"waypoint",
+				"list",
+				"--namespace",
+				nsConfig.Name(),
+			})
+			for _, sa := range saSet {
+				if !strings.Contains(output, sa) {
+					t.Fatalf("expect to find %s in output: %s", sa, output)
+				}
+			}
+
+			output, _ = istioctl.NewOrFail(t, t, istioctl.Config{}).InvokeOrFail(t, []string{
+				"x",
+				"waypoint",
+				"list",
+				"-A",
+			})
+			for _, sa := range saSet {
+				if !strings.Contains(output, sa) {
+					t.Fatalf("expect to find %s in output: %s", sa, output)
+				}
+			}
+
+			istioctl.NewOrFail(t, t, istioctl.Config{}).InvokeOrFail(t, []string{
+				"x",
+				"waypoint",
+				"-n",
+				nsConfig.Name(),
+				"delete",
+			})
+			retry.UntilSuccessOrFail(t, func() error {
+				fetch := kubetest.NewPodFetch(t.AllClusters()[0], nsConfig.Name(), constants.GatewayNameLabel+"="+"namespace")
```

Comment:
this fetch function is too repetitive and has a pattern. The only difference is the selector label. Maybe turn it into a function so we avoid code duplication?